### ISSUE TITLE
The config parser should throw error if there are issues

### ISSUE THREAD

**Issue by mmahmoudian** [flameshot-org/flameshot#1740]
## The issue

We have got multiple issues reported that Flameshot does not behave as expected and it has been narrowed down to mistakes in config file. the most recent issue was #1739.

## Suggestion

The part of the code that reads and parses the config file should:
1. throw error (in notification, in the log, and also as a layer to when the `flameshot gui` opens)
2. falls back to some defaults until the next trigger of reading the config

**Comment by veracioux** [comment]
I propose that we check the config file:
- at each launch of the daemon
- each time the config file changes a la `QFileSystemWatcher`

These are all the checks I could think of (I invite you to contribute more):
# Syntax checks are handled by `QSettings` already
Errors are indicated by a status flag.
# Perform semantic checks on the config data
Are numbers, dates, file extensions, etc correctly formatted.
# Check if there are unrecognized entries
If we add this check, I think it should be on demand by the user. Extra
entries in the config are not a big problem. The only thing they can really
do is clog up the config. But even this is unrealistic to expect.
# Check if there are multiple instances of an entry
I don't know how we would implement this tbh.

We should notify the user that there was an error using the means that
@mmahmoudian mentioned and then attempt to fix the error.

- Should we ask for confirmation before fixing it?
- Should we always show what the error was before fixing it?
- What errors can be fixed at all?

I have seen that spdlog is a dependency, but haven't seen anything useful
done with it. Do you use it for anything yet?

**Comment by mmahmoudian** [comment]
@veracioux 

> I propose that we check the config file:
> - at each launch of the daemon
> - each time the config file changes a la `QFileSystemWatcher`

I agree with your first bullet point, but the second one is overkill imho as [there will be unnecessary overhead load](https://doc.qt.io/qt-5/qfilesystemwatcher.html#details). I think a better alternative would be to run it whenever the daemon is about to perform an action  (basically all subcommands of `flameshot`, plus the launcher)

> **Perform semantic checks on the config data**
> Are numbers, dates, file extensions, etc correctly formatted.

As a related, but off-topic comment: I think it would be nice to add comment lines above each item in the `flameshot.ini` to guide the user.

> **Check if there are unrecognized entries**
> If we add this check, I think it should be on demand by the user.

I didn't understand this.

> **Check if there are multiple instances of an entry**
> I don't know how we would implement this tbh.

Disclaimer: I don't know C++ and QT good enough to have solid suggestions.
That said, this is actually very easy. There are two approaches that different software use:
1. The latter over-write the former. So if you have multiple instances of the same item in a specific section, the one with the largest line number would be the one used.
2. They produce error. This is typically done by either the library that reads ini, TOML, or YAML, or by the programmer through checking duplicates.

Either way we plan to go with (my suggestion is the latter as it will remove confusion), I can help in algorithmic procedure.

> ... and then attempt to fix the error automatically.

As much as I like automagical error correction, I personally don't think this is necessary in this case. We should just inform user about the error and possibly pointing them to the correct documentation (e.g web page, default config, ...).

> I have seen that spdlog is a dependency, but haven't seen anything useful done with it. Do you use it for anything yet?

I remember that @borgmanJeremy was working on some logging and error-reporting, but I don't know the status of the progress. I believe we should have a good logging method (opt-in by default, i.e: disabled by default).

Regarding how to show the user the config errors, I suggest all the following together:
1. notification
2. changing the tray icon (overlay a :no_entry_sign: or :interrobang: to the Flameshot logo)
3. write an error log file in the same path as the config file (e.g `~/.config/flameshot/flameshot.ini_error.txt`) should include the date and time
4. show some sort of error in the config gui (e.g a new tab that has only a text area to show the error(s) and a button for "Check again")
5. overlay to the `flameshot gui` (but the user should be able to close it and still take the screenshot as normal)
    - if the config file changed while the daemon was running, the new changes will not get applied and user will use the old loaded value of the config file
    - if the config was changed and then the daemon was executed, then it should fallback to the default config

Do you think this is over-complicated?

**Comment by veracioux** [comment]
> > * each time the config file changes a la `QFileSystemWatcher`

Perhaps you are right about the overhead. I think your proposition is good.

> As a related, but off-topic comment: I think it would be nice to add comment lines above each item in the `flameshot.ini` to guide the user.

I don't think this is necessary. IMO the options already have pretty descriptive names. And if someone wonders what an option does they can open the graphical menu to check. Besides, I don't know if QSettings lends itself well to this feature.
> > **Check if there are unrecognized entries**
> > If we add this check, I think it should be on demand by the user.
>
> I didn't understand this.

I meant to check if someone adds a config option that isn't recognized by flameshot. For example if someone makes a typo and writes `showHepl=true`.
> 
> > **Check if there are multiple instances of an entry**
> > I don't know how we would implement this tbh.

I meant that I'm not sure how we will work it into the current implementation that uses `QSettings`. I will probably have to refresh my knowledge of that class. But your comments are still useful, and as of now I am not sure which approach I prefer.

> As much as I like automagical error correction, I personally don't think this is necessary in this case. We should just inform user about the error and possibly pointing them to the correct documentation (e.g web page, default config, ...).

Noted, but I think this should be left undecided until I start working on the code. If the error correction proves easy, why not add it.
 
> Regarding how to show the user the config errors, I suggest all the following together:
> 
> 1. notification

Sure.

> 2. changing the tray icon (overlay a no_entry_sign or interrobang to the Flameshot logo)

I have a feeling this could cause some headaches implementation-wise. But can't say for sure until I try.

> 3. write an error log file in the same path as the config file (e.g `~/.config/flameshot/flameshot.ini_error.txt`) should include the date and time

Maybe another location, probably `~/.local/share/...`. This seems more standard according to my experience.

> 4. show some sort of error in the config gui (e.g a new tab that has only a text area to show the error(s) and a button for "Check again")

I agree.

> 5. overlay to the `flameshot gui` (but the user should be able to close it and still take the screenshot as normal)

I'm not sure this is necessary. If we do add this, I think it should be a standard warning dialog, and not an overlay. And it should be displayed before the screen gets taken over by flameshot.

> * if the config file changed while the daemon was running, the new changes will not get applied and user will use the old loaded value of the config file

Do you mean only if the new config has an error, or in general?

> * if the config was changed and then the daemon was executed, then it should fallback to the default config

Fine by me.

**Comment by mmahmoudian** [comment]
@veracioux 

> > As a related, but off-topic comment: I think it would be nice to add comment lines above each item in the `flameshot.ini` to guide the user.
> 
> I don't think this is necessary. IMO the options already have pretty descriptive names.

I agree that the names are descriptive enough. What I meant was to guide the user what type of input is acceptable. i.e `true`/`false`, hexadecimal color code, numeric, there is no `Enter` but rather `Return`, Ctrl, Shift, Del, ... should be capitalized, etc.

> If the error correction proves easy, why not add it.

I agree that if it is not easy, we should not even go to that direction. But if it was easy and we wanted to automate it, we should warn user about the change and possibly make it on-demand. Many do version control their dotfiles and I personally hate it when a software change the dotfile without my permission and knowledge (e.g [Konsole](https://konsole.kde.org/), [screenkey](https://gitlab.com/screenkey/screenkey/)) because it will ruin the `git status`. But I would be happy if they let me to choose if the autocorrection should happen.

> I have a feeling this could cause some headaches implementation-wise. But can't say for sure until I try.

Definitely go with the less painful approach 👍🏼 . Considering that users can now disable the notifications, we need to alert the user about bad config without being too much disruptive to their daily workflow and routine.



> > overlay to the flameshot gui (but the user should be able to close it and still take the screenshot as normal)
> 
> I'm not sure this is necessary. If we do add this, I think it should be a standard warning dialog, and not an overlay. And it should be displayed before the screen gets taken over by flameshot.

What I meant is perhaps what you explained. I meant something like the update message we show user in `flameshot gui`.

> > * if the config file changed while the daemon was running, the new changes will not get applied and user will use the old loaded value of the config file
> 
> Do you mean only if the new config has an error, or in general?

I meant if and only if the new config has error.

Btw, thanks for the time you are putting into this and discussing the details. I highly appreciate this 🍻

**Comment by veracioux** [comment]
Alright, first draft ready - and it is a hefty one: #1859. Sorry it took me so long, had some trouble with QSettings.

Please read the notes in the PR as well.

## Amendments
I have done a few things differently than agreed upon, either because it was
easier to implement or came more naturally.
1. I have added a file watcher.
    But I have implemented it to have as little  overhead as possible - and I haven't observed any performance issues.
2. About your 5th point
     If the changes do not not 
3. Config comments should be left only in `flameshot.example.ini`.
    I don't want to have to bypass the behavior of `QSettings` in order to add the comments in the live config file.
4. See **Notes about Qt's QSettings**

If you disagree with these, I will be happy to implement them the way we  agreed upon.

## Open questions

- What do we do about paths in the config?
  Do we consider a non-existing or non-directory path to be an error? What if the default path (~/Pictures) does not exist? For example I do not have ~/Pictures on my system. Should we fall back to `~` in that case? Can we rely that `~` exists?
- What to do with `useJpgForClipboard` which is disabled on MacOS in the code?
    Do we consider it an error if it is disabled for a given platform, or do we just ignore the result?
- It has become apparent to me that we shoud have a way for the user to find out where exactly config error is. I think we could be flooded with issue reports and user frustrations if we don't implement this. We do not need to automatically correct the errors (also a bit of laziness talking).  But I think a way to find out which config settings are erroneous is a must. Among other things, a CLI option to the `config` subcommand would be the most elegant solution. On the other hand, this would neglect Windows users. Thoughts?

## Notes about Qt's QSettings

Please read this even if you are not a C++ developer to be informed (you don't have to follow the links).
There is a lot of stuff that QSettings does, that we don't have control over. I don't think this is a big deal and we can let that pass, because I think it is acceptable and changing the behavior is a pain...
- Format errors are resolved automatically
   I have observed only the following types of format errors: broken section (e.g. `[Settings`, `Settings]`, `[[Settings]`, etc.); and invalid quotation of values.
   They are only a minor nuissance if you version control your config, but you should be careful not to make these basic mistakes anyway. They are fairly easy to spot, and they can't cause any breakage.
- We can't detect duplicate entries (e.g. file containing `showHelp=true` and `showHelp=false` simultaneously).
QSettings automatically removes all entries but the last one. Well, you could change this if you parse the file yourself or possibly by implementing [QSettings::registerFormat](https://doc.qt.io/qt-5/qsettings.html#registerFormat). I would like to avoid the former and am utterly unfamiliar with the latter (and the documentation seems scarce) so I would let this pass.
- It also sorts the config entries every time you try to write a setting, but we can't do anything about that (that I know of).
- Even trying to detect the errors in the config is problematic because `QSettings` has a weird behavior. It only reports an error the first time you ask for it. Every time you try to write a setting, it fixes all the format errors in the config. This has been reported as a [bug](https://bugreports.qt.io/browse/QTBUG-8346?jql=status%20%3D%20Closed%20AND%20text%20~%20%22qsettings%20status%22) to Qt but the issue was closed with a remark that this is expected behavior, without much elaboration.
- @borgmanJeremy If it's not too much trouble, I would like you to grep the diff and find any TODO entries I added and give me your feedback: `git diff  master..config_checker | grep '^+.*TODO'`

**Comment by mmahmoudian** [comment]
> What do we do about paths in the config?
> Do we consider a non-existing or non-directory path to be an error? What if the default path (~/Pictures) does not exist? For example I do not have ~/Pictures on my system. Should we fall back to ~ in that case? Can we rely that ~ exists?

Yes I think if the path does not exists it should throw an error. The fallback for example can be `~/Pictures`, `~`, `/tmp` in this order.

> What to do with useJpgForClipboard which is disabled on MacOS in the code?
> Do we consider it an error if it is disabled for a given platform, or do we just ignore the result?

I believe it should throw error.

> It has become apparent to me that we shoud have a way for the user to find out where exactly config error is. I think we could be flooded with issue reports and user frustrations if we don't implement this. We do not need to automatically correct the errors (also a bit of laziness talking). But I think a way to find out which config settings are erroneous is a must. Among other things, a CLI option to the config subcommand would be the most elegant solution. On the other hand, this would neglect Windows users. Thoughts?

I agree. we should **not** automatically correct anything. In my opinion, we can create a `errorLog.txt` file in the config folder right beside the config file. Additionally, we should also send the errors to stdout of the daemon as well.

**Comment by mmahmoudian** [comment]
## Notes about Qt's QSettings

> Format errors are resolved automatically

Fair enough

> We can't detect duplicate entries

Would be fine, but I think users can be confused if there are two instances and they miss seeing one, then they think there is a bug.

> It also sorts the config entries every time you try to write a setting, but we can't do anything about that 

This is fine but if someone version-control the config, they will get annoyed a bit until they commit the sorted version

> Even trying to detect the errors in the config is problematic because QSettings has a weird behavior.

Then nothing can be done. Right?

**Comment by veracioux** [comment]
@mmahmoudian I would postpone logging for the time being. My understanding from looking at the code is that logging is not implemented. I could implement some makeshift solution just for the purposes of the config checker, but I think it is better to wait for it to be implemented properly. Stdout is also a bit problematic, but I think I could implement it.

> Among other things, a CLI option to the config subcommand would be the most elegant solution.

For example `flameshot config --check` that prints all the settings that are wrong.

> This is fine but if someone version-control the config, they will get annoyed a bit until they commit the sorted version

That's a bummer, but I don't think we can't do anything about it without turning the code upside down. The best we can do is update the example config so that everything is sorted (if it isn't already).

> Would be fine, but I think users can be confused if there are two instances and they miss seeing one, then they think there is a bug.

It would annoy me too. But I'm afraid we can't do anything.

>> Even trying to detect the errors in the config is problematic because QSettings has a weird behavior. 

> Then nothing can be done. Right?

Afraid so.

**Comment by borgmanJeremy** [comment]
For logging I was planning to use spdlog, I kind of just add it in as I go. We can instead use the QT logging framework.
